third_party/move/move-vm/runtime/src/storage/ty_layout_converter.rs

```diff
@@ -185,229 +274,185 @@ pub(crate) trait LayoutConverterBase {
+        count: &mut u64,
+        depth: u64,
+    ) -> PartialVMResult<(Vec<MoveTypeLayout>, bool)> {
+        let mut contains_delayed_fields = false;
+        let layouts = tys
+            .iter()
+            .map(|ty| {
+                let (layout, ty_contains_delayed_fields) = self
+                    .type_to_type_layout_impl::<ANNOTATED>(
+                        gas_meter,
+                        traversal_context,
+                        ty,
+                        count,
+                        depth,
+                    )?;
+                contains_delayed_fields |= ty_contains_delayed_fields;
+                Ok(layout)
+            })
+            .collect::<PartialVMResult<Vec<_>>>()?;
+        Ok((layouts, contains_delayed_fields))
+    }
+
+    /// Converts a struct type into the corresponding layout. If a decorated layout is required
+    /// (annotated flag is set), field names are added to the layout. For delayed fields, marks
+    /// layout as containing delayed fields.
+    // TODO(lazy-loading):
+    //   We should add struct cyclic checks here as well, but this has to be done carefully because
+    //     struct A<T> {
+    //       x: B<B<u8>>,
+    //     }
+    //  is ok, even though x's type uses cycles between Bs.
+    fn struct_to_type_layout<const ANNOTATED: bool>(
         &self,
-        struct_name_idx: StructNameIndex,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut impl ModuleTraversalContext,
+        idx: &StructNameIndex,
         ty_args: &[Type],
         count: &mut u64,
         depth: u64,
     ) -> PartialVMResult<(MoveTypeLayout, bool)> {
-        let struct_type = self.fetch_struct_ty_by_idx(struct_name_idx)?;
-
-        let mut has_identifier_mappings = false;
+        use MoveTypeLayout as L;
 
-        let layout = match &struct_type.layout {
-            StructLayout::Single(fields) => {
-                // Some types can have fields which are lifted at serialization or deserialization
-                // times. Right now these are Aggregator and AggregatorSnapshot.
-                let maybe_mapping = self.get_identifier_mapping_kind(struct_name_idx)?;
-                let field_tys = fields
-                    .iter()
-                    .map(|(_, ty)| {
-                        self.vm_config()
-                            .ty_builder
-                            .create_ty_with_subst(ty, ty_args)
-                    })
-                    .collect::<PartialVMResult<Vec<_>>>()?;
-                let (mut field_layouts, field_has_identifier_mappings): (
-                    Vec<MoveTypeLayout>,
-                    Vec<bool>,
-                ) = field_tys
-                    .iter()
-                    .map(|ty| self.type_to_type_layout_impl(ty, count, depth))
-                    .collect::<PartialVMResult<Vec<_>>>()?
-                    .into_iter()
-                    .unzip();
+        let apply_subst = |tys: &[(Identifier, Type)]| {
```

Comment:
both use cases seem to be inside type_to_type_layouts, so we might be able to extract that whole call including the implementation of this closure into a helper method? (type_to_type_layouts_with_substituted_fields or a better name:)

Reply:
Extracted into a method. `types_to_type_layouts` is still used for function layouts though?